python/dgl/backend/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,356 @@
+                for idx, data in emb._trace:
+                    grad = data.grad.data
+                    device = grad.device
+                    idx_dtype = idx.dtype
+                    grad_dtype = grad.dtype
+                    grad_dim = grad.shape[1]
+
+                    if self._world_size > 0:
+                        if emb_name not in self._shared_cache:
+                            self._shared_cache[emb_name] = {}
+
+                        for i in range(self._world_size):
+                            start = i * range_size
+                            end = (i + 1) * range_size \
+                                if (i + 1) * range_size < num_embeddings \
+                                else num_embeddings
+                            if i == 0:
+                                mask = idx < end
+                            elif i + 1 == self._world_size:
+                                mask = idx >= start
+                            else:
+                                mask = th.logical_and((idx >= start),(idx < end))
+                            idx_i = idx[mask]
+                            grad_i = grad[mask]
+
+                            if i == self._rank:
+                                shared_emb[emb_name][0].append(idx_i)
+                                shared_emb[emb_name][1].append(grad_i)
+                            else:
+                                # currently nccl does not support Alltoallv operation
+                                # we need to use CPU shared memory to share gradient across processes
+                                idx_i = idx_i.to(th.device('cpu'))
+                                grad_i = grad_i.to(th.device('cpu'))
+                                idx_shmem_name = 'idx_{}_{}_{}'.format(emb_name, self._rank, i)
+                                grad_shmem_name = 'grad_{}_{}_{}'.format(emb_name, self._rank, i)
+
+                                if idx_shmem_name not in self._shared_cache[emb_name] or \
+                                    self._shared_cache[emb_name][idx_shmem_name].shape[0] < idx_i.shape[0]:
+                                    idx_shmem = create_shared_mem_array(idx_shmem_name,
+                                        (idx_i.shape[0] * 2 + 2,), idx_dtype) # in case idx_i.shape[0] is 0
+                                    grad_shmem = create_shared_mem_array(grad_shmem_name,
+                                        (idx_i.shape[0] * 2 + 2, grad_dim), grad_dtype)
+                                    self._shared_cache[emb_name][idx_shmem_name] = idx_shmem
+                                    self._shared_cache[emb_name][grad_shmem_name] = grad_shmem
+
+                                self._shared_cache[emb_name][idx_shmem_name][:idx_i.shape[0]] = idx_i
+                                self._shared_cache[emb_name][grad_shmem_name][:idx_i.shape[0]] = grad_i
+                                emb.store.set(idx_shmem_name, str(idx_i.shape[0]))
```

Comment:
is this a good way to send gradients to other processes? is it faster than using queue?

Reply:
in this way, you need to constantly create shared-memory and remove it. i'm wondering how expensive it is. i never tested it.

Reply:
I only create the shared-memory once and reuse it again and again.